### PR TITLE
feat: update api path and keep alive

### PR DESCRIPTION
**Pull request by lostbean**
## Description:
Update REST API path with `/api` prefix and add websocket ping/pong handler to keep connection alive.

## Is this change user facing?
NO

### CODE REVIEW

File: engine/server/engine/main.go [hunk 1]
```diff
@@ -411,14 +413,15 @@ func restApiServer(
 
 	// This is how you set up a basic Echo router
 	echoRouter := echo.New()
-	echoRouter.Use(echomiddleware.Logger())
+	echoApiRouter := echoRouter.Group("/api")
```
Comment (leoporoli) [inline engine/server/engine/main.go]: I think you can use this constant here `pathToApiGroup`

File: engine/server/engine/main.go [hunk 2]
```diff
@@ -411,14 +413,15 @@ func restApiServer(
...
+	echoApiRouter.Use(echomiddleware.Logger())
```
Comment (leoporoli) [inline engine/server/engine/main.go:417]: nice!

File: engine/server/engine/server/websocket_api_handler.go [hunk 3]
```diff
@@ -245,32 +274,43 @@ func streamStarlarkLogsWithHTTP[T any](ctx echo.Context, streamerPool streaming.
...
-		err := streamer.Consume(func(logline *api_type.ServiceLogs) error {
-			err := websocket.JSON.Send(ws, logline)
-			if err != nil {
-				return err
-			}
-			return nil
+func streamServiceLogsWithWebsocket(ctx echo.Context, cors cors.Cors, streamer streaming.ServiceLogStreamer) {
+	wsPump, err := NewWebsocketPump[api_type.ServiceLogs](ctx, cors)
+	if err != nil {
+		logrus.WithError(err).Error("Failed to start websocket connection")
```
Comment (leoporoli) [inline engine/server/engine/server/websocket_api_handler.go:264]: I think you could create a helper function here to dry a bit, it's the same used in `streamStarlarkLogsWithWebsocket`

File: engine/server/engine/server/websocket_api_handler.go [hunk 4]
```diff
@@ -245,32 +274,43 @@ func streamStarlarkLogsWithHTTP[T any](ctx echo.Context, streamerPool streaming.
...
+	defer wsPump.Close()
+	go wsPump.startPumping()
+
+	err = streamer.Consume(func(logline *api_type.ServiceLogs) error {
+		wsPump.pumpMessage(logline)
+		return nil
+	})
+
+	if err != nil {
+		logrus.WithError(err).WithFields(logrus.Fields{
```
Comment (leoporoli) [inline engine/server/engine/server/websocket_api_handler.go:281]: same for this error

File: engine/server/engine/server/websocket_api_handler.go [hunk 5]
```diff
@@ -300,3 +340,95 @@ func streamServiceLogsWithHTTP(ctx echo.Context, streamer streaming.ServiceLogSt
 		}
 	}
 }
+
+type WebsocketPump[T interface{}] struct {
```
Comment (leoporoli) [inline engine/server/engine/server/websocket_api_handler.go]: small suggestion, I think this object and methods could be on its file

File: engine/server/engine/server/websocket_api_handler.go [hunk 6]
```diff
@@ -300,3 +340,95 @@ func streamServiceLogsWithHTTP(ctx echo.Context, streamer streaming.ServiceLogSt
...
+
+func NewWebsocketPump[T interface{}](ctx echo.Context, cors cors.Cors) (*WebsocketPump[T], error) {
+	// nolint: exhaustruct
+	upgrader := websocket.Upgrader{
+		ReadBufferSize:  wsReadBufferSize,
+		WriteBufferSize: wsWriteBufferSize,
+		CheckOrigin:     cors.OriginAllowed,
+	}
+
+	conn, err := upgrader.Upgrade(ctx.Response(), ctx.Request(), nil)
```
Comment (leoporoli) [inline engine/server/engine/server/websocket_api_handler.go]: nice!

File: engine/server/engine/main.go (hunk 1 above)
Comment (lostbean) [reply engine/server/engine/main.go]: fixed. nice catch!

File: engine/server/engine/server/websocket_api_handler.go [hunk 7]
```diff
@@ -300,3 +340,95 @@ func streamServiceLogsWithHTTP(ctx echo.Context, streamer streaming.ServiceLogSt
...
+func (pump *WebsocketPump[T]) pumpResponseInfo(msg *api_type.ResponseInfo) {
+	pump.infoChan <- msg
+}
+
+func (pump *WebsocketPump[T]) pumpMessage(msg *T) {
+	pump.inputChan <- msg
+}
+
+func (pump *WebsocketPump[T]) Close() {
+	pump.cancelFunc()
```
Comment (leoporoli) [inline engine/server/engine/server/websocket_api_handler.go]: I think you can close the web socket here too and the channels, or is it necessary to let them open?

Comment (leoporoli) [reply engine/server/engine/server/websocket_api_handler.go]: and the ticker?

File: engine/server/engine/server/websocket_api_handler.go (hunk 4 above)
Comment (lostbean) [reply engine/server/engine/server/websocket_api_handler.go:281]: Here it's different, we are returning the error over the websocket channel. I can't use the same abstraction.

File: engine/server/engine/streaming/websocket_pump.go [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+
+	logrus.WithFields(logrus.Fields{
+		"pongWait":       pongWait,
+		"pingPeriod":     pingPeriod,
+		"maxMessageSize": maxMessageSize,
+	}).Debug("Started keep alive process for websocket connection.")
+
+	pump.websocket.SetReadLimit(maxMessageSize)
+	if err := pump.websocket.SetReadDeadline(time.Now().Add(pongWait)); err != nil {
+		logrus.WithError(err).Error("Failed to set Pong wait time")
```
Comment (leoporoli) [inline engine/server/engine/streaming/websocket_pump.go:78]: will the client receive and error if this happens?

File: engine/server/engine/server/websocket_api_handler.go (hunk 7 above)
Comment (lostbean) [reply engine/server/engine/server/websocket_api_handler.go]: If I do it here, we'll be closing it twice and it'll throw an error. The goroutine `StartPumping()` will always close it on exit (`defer`).

File: engine/server/engine/streaming/websocket_pump.go [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+
+func (pump WebsocketPump[T]) StartPumping() {
+	ticker := time.NewTicker(pingPeriod)
+	defer func() {
+		ticker.Stop()
+		pump.websocket.Close()
+	}()
+
+	logrus.WithFields(logrus.Fields{
+		"pongWait":       pongWait,
```
Comment (leoporoli) [inline engine/server/engine/streaming/websocket_pump.go:71]: nit: you could use constants for these keys

File: engine/server/engine/streaming/websocket_pump.go (hunk 8 above)
Comment (lostbean) [reply engine/server/engine/streaming/websocket_pump.go:78]: no, this is logged only the server side. if this fails, it's very likely the connection is broken.

File: engine/server/engine/streaming/websocket_pump.go (hunk 9 above)
Comment (lostbean) [reply engine/server/engine/streaming/websocket_pump.go:71]: what's the benefity in this case? it does seems an overkill, no?

**Review by leoporoli — APPROVED** [review]

File: engine/server/engine/server/websocket_api_handler.go (hunk 7 above)
Comment (leoporoli) [reply engine/server/engine/server/websocket_api_handler.go]: @lostbean makes sense, so the close channels could be included in the defer on exit? it is suggested to always close the channels when the writer ends the task

File: engine/server/engine/streaming/websocket_pump.go (hunk 9 above)
Comment (leoporoli) [reply engine/server/engine/streaming/websocket_pump.go:71]: probably is an over-engineering suggestion, I use to do it for me or someone else copy the code later both pieces have the same key and then if it is edited, will be edited in only one place, but.
Not necessary to do the change.

File: engine/server/engine/server/websocket_api_handler.go (hunk 7 above)
Comment (lostbean) [reply engine/server/engine/server/websocket_api_handler.go]: yes, I forgot to close the go channels!

Comment (lostbean) [reply engine/server/engine/server/websocket_api_handler.go]: done!

**Pull request merged**